Corona-Warn-App/src/main/java/de/rki/coronawarnapp/diagnosiskeys/download/KeyDownloadTool.kt

```diff
@@ -0,0 +1,69 @@
+package de.rki.coronawarnapp.diagnosiskeys.download
+
+import dagger.Reusable
+import de.rki.coronawarnapp.appconfig.KeyDownloadConfig
+import de.rki.coronawarnapp.diagnosiskeys.server.DiagnosisKeyServer
+import de.rki.coronawarnapp.diagnosiskeys.server.DownloadInfo
+import de.rki.coronawarnapp.diagnosiskeys.storage.CachedKey
+import de.rki.coronawarnapp.diagnosiskeys.storage.KeyCacheRepository
+import de.rki.coronawarnapp.diagnosiskeys.storage.legacy.LegacyKeyCacheMigration
+import de.rki.coronawarnapp.util.HashExtensions.hashToMD5
+import kotlinx.coroutines.withTimeout
+import timber.log.Timber
+import javax.inject.Inject
+
+@Reusable
+class KeyDownloadTool @Inject constructor(
+    private val legacyKeyCache: LegacyKeyCacheMigration,
+    private val keyServer: DiagnosisKeyServer,
+    private val keyCache: KeyCacheRepository
+) {
+    suspend fun downloadKeyFile(
+        cachedKey: CachedKey,
+        downloadConfig: KeyDownloadConfig
+    ): CachedKey = try {
+        val saveTo = cachedKey.path
+        val keyInfo = cachedKey.info
+
+        val preconditionHook: suspend (DownloadInfo) -> Boolean =
+            { downloadInfo ->
+                /**
+                 * To try legacy migration, we attempt to the etag as checksum.
+                 * Removing the quotes, the etag can represent the file's MD5 checksum.
+                 */
+                val etagAsChecksum = downloadInfo.etagWithoutQuotes
+                val continueDownload = !legacyKeyCache.tryMigration(etagAsChecksum, saveTo)
+                continueDownload // Continue download if no migration happened
+            }
+
+        val dlInfo = withTimeout(downloadConfig.individualDownloadTimeout.millis) {
```

Comment:
I assume `dl` stands for download? Everywhere else it's written out, so why not also do it here?